### DISCUSSION TITLE
pyMOR 2024.1.0

### DISCUSSION THREAD

**Discussion by sdrave** [pymor/pymor#2301] [Announcements]
We are proud to announce the release of pyMOR 2024.1!

The main new features are:

* vector fitting
* successive constraints method
* shifted Cholesky QR algorithm
* improved randomized range finder
* additional tutorials and MOR methods overview

The full release notes are available [here](https://docs.pymor.org/2024-1-0/release_notes/all.html#pymor-2024-1-july-8-2025).

pyMOR 2024.1 contains contributions by @maxbindhak  and @artpelling. See [here](https://github.com/pymor/pymor/blob/main/AUTHORS.md) for more details.

<hr /><em>This discussion was created from the release <a href='https://github.com/pymor/pymor/releases/tag/2024.1.0'>pyMOR 2024.1.0</a>.</em>

**Comment by cTatu** [comment]
Hello,

I wanted to try the shifted cholesky qr algorithm for a tall-skinny matrix of shape (802623, 270).
```python
import numpy as np
import pymor.algorithms.chol_qr
from pymor.basic import NumpyVectorSpace

if __name__ == '__main__':
    arr = np.load('snapshots_sol0.npy') # shape (802623, 270)

    m, n = arr.shape

    space = NumpyVectorSpace(n)

    var = space.from_numpy(arr)

    q, r = pymor.algorithms.chol_qr.shifted_chol_qr(var)

    orth = np.linalg.norm((q.T @ q) - np.eye(n)) / np.sqrt(n)
    residual = np.linalg.norm(arr - (q @ r)) / np.linalg.norm(arr)

    print(f'\tORTHOGONALITY: {orth} | RESIDUAL: {residual}')
```

But in the algorithm there was this error:
```
File "pymor/algorithms/chol_qr.py", line 85, in shifted_chol_qr
    B, X = np.split(A[offset:].inner(A, product=product), [offset], axis=1)
numpy.core._exceptions.MemoryError: Unable to allocate 4.69 TiB for an array with shape (802623, 802623) and data type float64
```

I was doing something wrong converting the numpy array to NumpyVectorSpace?

Thank you

**Reply by maxbindhak** [reply]
Hello,

try to convert your numpy array into a VectorArray using
```python
space = NumpyVectorSpace(m)
var = space.from_numpy(arr.T)
```

Since numpy stores the matrix in C major order by default the decision was made to consider the rows of a matrix as the columns/vectors. This might change in the future.

**Reply by cTatu** [reply]
Thank you for the quick response.
I did that and the algorithm is working but is not converging.
```
$ python3 ch.py
00:02 shifted_chol_qr: Iteration 1
00:02 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:02 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:06 shifted_chol_qr: Iteration 2
00:06 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:06 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:09 shifted_chol_qr: Iteration 3
00:09 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:09 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:11 shifted_chol_qr: Iteration 4
00:11 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:11 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:14 shifted_chol_qr: Iteration 5
00:14 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:14 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:17 shifted_chol_qr: Iteration 6
00:17 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:17 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:20 shifted_chol_qr: Iteration 7
00:20 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:20 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:23 shifted_chol_qr: Iteration 8
00:23 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:23 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:25 shifted_chol_qr: Iteration 9
00:25 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:25 |   shifted_chol_qr: Applying shift: 14514323.601256896
00:28 shifted_chol_qr: Iteration 10
00:28 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:28 |   shifted_chol_qr: Applying shift: 14514323.601256896

            ORTHOGONALITY: 1.0 | RESIDUAL: 8.643828142447564e-16
```

The shift should be far smaller right? I think is computed based on the `m,n` shape of the array but since is transposed it might be wrongly calculated?

I uploaded the input numpy array [here](https://zenodo.org/records/12917365/files/snapshots_sol0.npy?download=1) if you want to try it out.

Thank you very much

**Comment by artpelling** [comment]
Hi @cTatu ,

its nice to see that other people are using the algorithm :)
The problem your facing is that pyMOR uses a transposed (in comparison to NumPy) formulation for `VectorArrays`. So you will always need to transpose when converting between NumPy and pyMOR and vice versa.
I noticed that your snippet also had a problem later on. Here is an updated snippet that should run and contains some more suggestions:
``` python
#!/usr/bin/env python3

import numpy as np
from pymor.algorithms.chol_qr import shifted_chol_qr
from pymor.basic import NumpyVectorSpace


# consider setting this for an adaptive algorithm
orth_tol = None

if __name__ == '__main__':
    arr = np.load('snapshots_sol0.npy') # shape (802623, 270)

    m, n = arr.shape

    space = NumpyVectorSpace(m)

    # pyMOR uses a transposed notation, so converting from and to numpy needs a transpose
    V = space.from_numpy(arr.T)

    Q, R = shifted_chol_qr(V, orth_tol=orth_tol)

    # you can either use the NumpyArrays in pyMOR syntax
    orth = np.linalg.norm(Q.gramian() - np.eye(n)) / np.sqrt(n)
    residual = np.linalg.norm((V - Q.lincomb(R.T)).norm()) / np.linalg.norm(V.norm())  # VectorArray.norm() only calculates the norm of each element
    print(f'\tORTHOGONALITY: {orth} | RESIDUAL: {residual}')

    # or convert to numpy and then calculacte the residuals
    q, r = Q.to_numpy().T, R
    orth = np.linalg.norm((q.T @ q) - np.eye(n)) / np.sqrt(n)
    residual = np.linalg.norm(arr - (q @ r)) / np.linalg.norm(arr)
    print(f'\tORTHOGONALITY: {orth} | RESIDUAL: {residual}')
```

**Reply by cTatu** [reply]
Hello, thank you for the response.

I wasn't aware of all these details and your clarification was very helpful. I just discovered this library some hours ago and wanted to try a 'professional' implementation of the shifted cholesky qr algorithm. I've tried implementing this algorithm and others like [rQR-CholeskyQR (Algorithm 2.6)](https://arxiv.org/pdf/2111.11148) for factorizing my ill-conditioned matrix. None of them managed to successfully factorize it. That's why I tried pyROM. 
With your suggestion the convergence is still not achieved:
```
00:01 shifted_chol_qr: Iteration 1
00:01 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:01 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:06 |   shifted_chol_qr: Residual = 16.397452559982316
00:06 shifted_chol_qr: Iteration 2
00:06 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:06 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:08 |   shifted_chol_qr: Residual = 16.431676725153487
00:08 shifted_chol_qr: Iteration 3
00:08 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:08 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:11 |   shifted_chol_qr: Residual = 16.431676725154983
00:11 shifted_chol_qr: Iteration 4
00:11 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:11 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:14 |   shifted_chol_qr: Residual = 16.431676725154983
00:14 shifted_chol_qr: Iteration 5
00:14 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:14 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:17 |   shifted_chol_qr: Residual = 16.431676725154983
00:17 shifted_chol_qr: Iteration 6
00:17 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:17 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:20 |   shifted_chol_qr: Residual = 16.431676725154983
00:20 shifted_chol_qr: Iteration 7
00:20 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:20 |   shifted_chol_qr: Applying shift: 43146407228.93191
00:23 |   shifted_chol_qr: Residual = 16.431676725154983
00:23 shifted_chol_qr: Iteration 8
00:23 |   |WARNING|shifted_chol_qr: Cholesky factorization broke down! Matrix is ill-conditioned.
00:23 |   shifted_chol_qr: Applying shift: 43146407228.93191
```

I don't know if the shift is supposed to be that high, but the residual is not improving. Maybe this matrix is so bad conditioned that is impossible to factorize? `np.linalg.qr()` does a good job but I need to use some sort of CholeskyQR algorithm.
Do you have any ideas about how to resolve this?
I uploaded the input numpy array [here](https://zenodo.org/records/12917365/files/snapshots_sol0.npy?download=1) in case you want to analyze it.

Thank you very much

**Reply by artpelling** [reply]
I tested your example and it also does not converge for me. Unfortunately, the algorithm has trouble dealing with `VectorArrays` that contain linear dependent vectors;  resulting in a loss of orthogonality. This is an effect that we also experienced during testing but is probably related to the algorithm rather than the implementation.

It would be nice to have a rank-revealing version of this algorithm that can remove these vectors. We suspect that this is possible by using a [pivoted Cholesky decomposition](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=10579138
)~, but so far, no work that we know of has been published in this regard.~ 
If you cannot prevent your data being linear dependent, you could use `gram_schmidt` as
```python
from pymor.algorithms.gram_schmidt import gram_schmidt
Q, R = gram_schmidt(V, return_R=True)

# or, if you do not want to remove vectors
Q, R = gram_schmidt(V, atol=0, rtol=0, return_R=True)
```
which gives me
```
ORTHOGONALITY: 4.283964822466486e-15 | RESIDUAL: 4.587786334030428e-16
```

**Reply by artpelling** [reply]
> This is an effect that we also experienced during testing but is probably related to the algorithm rather than the implementation.


That being said, the results are still a little bit bewildering to me. Your matrix has a condition number of  `4*10^29` which is large but we have successfully used the algorithm with worse conditioned matrices during development. Furthermore, `gram_schmidt` achieves a low residual without the need to remove vectors. @pmli, @sdrave what do you think about this?

**Reply by cTatu** [reply]
Hi, thank you for the analysis.

This matrix is a representative example of a much larger matrix containing CFD simulations from which we need to compute a ROM (Reduced Order Model) using RandomSVD in a distributed environment (splitting the matrix into chunks). Inside the RandomSVD algorithm we've been using the TSQR algorithm, but its orthogonality deteriorates when the chunks don't contain all the matrix columns. That's why I'm trying to find an alternative algorithm. 

I found [this free access paper](https://arxiv.org/pdf/2210.09953#:~:text=The%20rank%2Drevealing%20randomized%20Cholesky,when%20X%20is%20rank%2Ddeficient.) about **Rank-revealing randomized Cholesky QR** (Algorithm 7). My results aren't great but I'm not confident that my implementation is correct. What do you think about this method?

Thank you very much

**Reply by maxbindhak** [reply]
Hi,

i tested out another variant of the shifted Cholesky QR, which you can find here  #2304 .

With that version i got the following results:

```bash
Loaded: path='./snapshots_sol0.npy' - m=802623 - n=270 - A_norm=5235654.227116746
.
.
.
t=17.567615299s
LOO=2.04221989180589e-14 # loss of orthogonality
RRR=1.6529013506232122e-15 # relative reconstruction residual
RCR=1.987037910062511e-14 # relative Cholesky residual
SubAng=2.5404007546272595e-06 # subspace angle between A and Q
```

using

```python
import numpy as np
from scipy.linalg import eigh, subspace_angles

from pymor.vectorarrays.numpy import NumpyVectorSpace
from pymor.algorithms.chol_qr import shifted_chol_qr
from time import perf_counter_ns

def _returnNanOnException(func):
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except Exception as e:
            return np.nan
    return wrapper

def _norm(X) -> float:
    return np.linalg.norm(X, ord=2)

def _normSym(X) -> float:
    n = X.shape[0]
    return np.abs(eigh(X, eigvals_only=True, subset_by_index=[n-1,n-1], driver='evr')[0])

@_returnNanOnException
def _lossOfOrthogonality(Q):
    return _normSym(np.eye(Q.shape[1]) - Q.T.dot(Q))

@_returnNanOnException
def _relativeResidual(A, Q, R, A_norm):
    return _norm(A - Q.dot(R)) / A_norm

@_returnNanOnException
def _relativeCholeskyResidual(A, R, A_norm):
    return _normSym(A.T.dot(A) - R.T.dot(R)) / A_norm**2

@_returnNanOnException
def _subspaceAngles(A, Q):
    return subspace_angles(A, Q)[0]


path = './snapshots_sol0.npy'
A = np.load(path)
m,n = A.shape
A_norm = _norm(A)
print(f'Loaded: {path=} - {m=} - {n=} - {A_norm=}')
A_ = NumpyVectorSpace(A.shape[0]).from_numpy(A.T)

t = -perf_counter_ns()
Q, R = shifted_chol_qr(A_, maxiter=10, orth_tol=1e-13, copy=False)
t += perf_counter_ns()
t /= 10**9

Q = Q.to_numpy().T

print(f'{t=}s')
LOO = _lossOfOrthogonality(Q)
print(f'{LOO=}')
RRR = _relativeResidual(A, Q, R, A_norm)
print(f'{RRR=}')
RCR = _relativeCholeskyResidual(A, R, A_norm)
print(f'{RCR=}')
SubAng = _subspaceAngles(A, Q)
print(f'{SubAng=}')
```

Since `scipy.sparse.linalg.eigsh` was used for the norm calculation, which uses a random initial vector by default,  your results might be different.

**Reply by artpelling** [reply]
That's a great point @maxbindhak. I was not sure what recomputing the shift does mathematically, but in light of https://arxiv.org/pdf/2111.11148, it can be simply seen as a preconditioner. So recomputing the shift makes a lot of sense. Based on the condition number of the provided matrix, about three iterations should be necessary in order for `sCholQR2` to give good results.

Here is a snippet I made starting from @maxbindhak suggestion but using `pymor/main` instead of #2304 if you want to work with our latest release. It will throw a few unecessary warnings but works with five iterations total:
```python
#!/usr/bin/env python3

import numpy as np
from pymor.core.logger import getLogger
from pymor.algorithms.chol_qr import shifted_chol_qr
from pymor.basic import NumpyVectorSpace


copy = False
orth_tol = None
maxiter = 5

if __name__ == '__main__':
    arr = np.load('snapshots_sol0.npy') # shape (802623, 270)

    m, n = arr.shape

    space = NumpyVectorSpace(m)

    # pyMOR uses a transposed notation, so converting from and to numpy needs a transpose
    V = space.from_numpy(arr.T)
    Q, R = V.copy(), np.eye(n)
    logger = getLogger('pymor.algorithms.chol_qr.shifted_chol_qr')
    for i in range(maxiter):
        with logger.block(f'Outer iteration {i+1}'):
            Q, Ri = shifted_chol_qr(Q, orth_tol=orth_tol, maxiter=1, copy=copy)
            R = Ri @ R

            orth = np.linalg.norm(Q.gramian() - np.eye(n)) / np.sqrt(n)
            residual = np.linalg.norm((V - Q.lincomb(R.T)).norm()) / np.linalg.norm(V.norm())  # VectorArray.norm() only calculates the norm of each element
            print(f'\tORTHOGONALITY: {orth} | RESIDUAL: {residual}')

```

I think we should definitely add the recalculation of shifts in order to make the algorithm more versatile and stable!

**Reply by sdrave** [reply]
Maybe we can move further discussion to #2304, to make it more visible? If the change significantly improves the algorithm, we should check if this has somehow been published before. If not, we might want to conduct further research ..

@cTatu, by 'distributed' you mean that the data comes live from an MPI distributed simulation? Otherwise, depending on your use case, you might also want to look at the HAPOD algorithm (disclaimer: my own work), which is trivial to parallelize, very robust and can easily handle rank-deficient data. The original algorithm formulation neglects the right-singular vectors, but these can be easily computed as well, if needed.

**Reply by cTatu** [reply]
Hello @sdrave sorry I got distracted.
The HAPOD algorithm is very interesting, I think it may be exactly what we need for the project. I'm a little bit familiar with SVD and PCA but I still have to wrap my head around Proper Orthogonal Decomposition (which I think is what we really need to do).
**Short answer to your question**: No, we don't use MPI and all data is already stored in disk.
**Long answer**:
The data is from large scale CFD simulations dumped to disk. Our part consists in implementing RandomSVD using [COMPSs](https://compss-doc.readthedocs.io/en/latest/index.html) (a tasked based distributed framework very similar to dask) in order to compute a truncated matrix of this tall and skinny matrix of shape (4656929, 9870). The computation is distributed by splitting the matrix in 2D blocks. 
Internally our RandomSVD algorithm uses TSQR algorithm for Q orthogonalization. But TSQR returns bad orthogonality results when the matrix is split in many chunks. That's why I wanted to replace TSQR with something more robust. I will take a look at HAPOD, it could be perfect because now I remember them saying that they only need the left singular vectors. The project is almost finished so there won't be big changes but at least it could be included in [dislib](https://dislib.readthedocs.io/en/latest/)
